spot-client/src/common/logger/post-to-endpoint.js

```diff
@@ -54,19 +96,76 @@ export default class PostToEndpoint {
             }
         }, []);
 
-        return fetch(
-            this._endpointUrl,
-            {
-                body: JSON.stringify({
-                    events: jsonEvents,
-                    'device_id': this._deviceId
-                }),
-                headers: {
-                    'Content-Type': 'application/json'
-                },
-                method: 'POST',
-                mode: 'cors'
-            }
-        );
+        this._requestsQueue.push(
+            new PostLogsRequest(/* attempt */ 1, jsonEvents));
+
+        // Make sure that there's only 1 request on the flight by starting
+        // the reaction chain only if there's 1 request in the queue (the one
+        // added above).
+        if (this._requestsQueue.length === 1) {
+            this._sendLogs();
+        }
+    }
+
+    /**
+     * Tries to send first log request from the requests queue. In case of
+     * success it will keep taking requests of the queue until it's empty.
+     * On failure it will retry the request with a delay until
+     * the {@link PostToEndpoint#MAX_ATTEMPTS} is reached.
+     *
+     * @returns {void}
+     * @private
+     */
+    _sendLogs() {
+        const request = this._requestsQueue.shift();
+
+        if (request) {
+            const { attempt, events } = request;
+
+            fetch(
+                this._endpointUrl,
+                {
+                    body: JSON.stringify({
+                        events,
+                        'device_id': this._deviceId
+                    }),
+                    headers: {
+                        'Content-Type': 'application/json'
+                    },
+                    method: 'POST',
+                    mode: 'cors'
+                }
+            ).then(response => {
+                const { ok, status } = response;
+
+                if (!ok && status >= 500 && status < 600) {
+                    // Retry on server errors and network failures which
+                    // is a regular fetch reject case.
+                    throw new Error(`Request error - status: ${status}`);
+                } else if (ok) {
+                    // Try sending the next request from the queue
+                    this._sendLogs();
+                } else {
+                    // eslint-disable-next-line no-console
+                    console.warn(`Dropping log request, status: ${status}`);
```

Comment:
Should you still call send logs here in case there’s something queued?

Reply:
yes